pkg/kv/kvserver/liveness/liveness.go

```diff
@@ -700,6 +687,8 @@ func (nl *NodeLiveness) Start(ctx context.Context) {
 					return nil
 				}); err != nil {
 				log.Warningf(ctx, heartbeatFailureLogFormat, err)
+			} else if nl.onSelfHeartbeat != nil {
+				nl.onSelfHeartbeat(ctx)
```

Comment:
This misses a bunch of cases. In particular, explicit `Heartbeat()` calls made during lease acquisitions.

We may want to move this into `heartbeatInternal` or something, or keep it in `updateLiveness` with the proper condition.

Reply:
> This misses a bunch of cases. In particular, explicit Heartbeat() calls made during lease acquisitions.

This was intentional, see commit message. There is no need for "every" heartbeat to bump this timestamp. But we do want the last up timestamp to rise along at the pace of the heartbeat loop, which this code achieves.

The last up timestamp is used only to inform the "node up" event that is emitted to the event log. There is no way to record the "true" last time the node is up anyway (what does that even mean?), so this mechanism keeps it simple and gets the job done.

Please let me know if you think I'm missing something here.

Reply:
This is definitely a bug before. The `onSelfHeartbeat` call should not be made for other nodes. Thanks for catching and fixing it!

Reply:
> The last up timestamp is used only to inform the "node up" event that is emitted to the event log.

Oh right, that thing. I'd be inclined to just remove that outright, but yeah, let's leave that for another day.